community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,21 +54,91 @@
+        return type(self) == type(other)
+
+class NodeActionResume(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """resume nodes via scontrol"""
+        log.info(f"{count} instances to resume ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_up")
+
+class NodeActionIdle(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """idle nodes"""
+        log.info(f"{count} nodes to idle ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=resume")
+
+class NodeActionRestart(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """start instances for nodes"""
+        log.info(f"{count} instances restarted ({hostlist})")
+        start_instances(nodes)
+
+class NodeActionDown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int, reason:str="Instance stopped/deleted") -> None:
+        """down nodes"""
+        log.info(
+            f"{count} nodes set down due to node status '{status.name}' ({hostlist})"
+        )
+        run(
+            f"{lookup().scontrol} update nodename={hostlist} state=down reason='{reason}'"
+        )
+
+class NodeActionPowerDown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """power_down node in slurm"""
+        log.info(f"{count} instances to power down ({hostlist})")
+        run(f"{lookup().scontrol} update nodename={hostlist} state=power_down")
+
+class NodeActionDelete(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """delete instances for nodes"""
+        log.info(f"{count} instances to delete ({hostlist})")
+        delete_instances(nodes)
+
+class NodeActionUnknown(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        """Error status, nodes shouldn't get in this status"""
+        log.error(f"{count} nodes have unexpected status: ({hostlist})")
+        first = next(iter(nodes))
+        state = lookup().slurm_node(first)
+        state = "{}+{}".format(state.base, "+".join(state.flags)) if state else "None"
+        inst = lookup().instance(first)
+        log.error(f"{first} state: {state}, instance status:{inst.status}")
+
+class NodeActionRestore(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        NodeActionIdle.apply(self, nodes, hostlist, count)
+
+class NodeActionPrempt(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        NodeActionDown.apply(self, nodes, hostlist, count, "Preempted instance")
+        NodeActionRestart.apply(self, nodes, hostlist, count)
+
+class NodeActionOrphan(NodeAction):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
```

Comment:
Why create duplicate actions?

Reply:
Did it for parity with `NodeStatus` initially, removing redundant actions.